sdk/filter/filter.go

```diff
@@ -2,43 +2,68 @@ package filter
 
 import "github.com/hypertrace/goagent/sdk"
 
-// Filter runs on request evaluates whether request should be
-// blocked
-type Filter struct {
-	URLEvaluators     []func(span sdk.Span, URL string) bool
-	HeadersEvaluators []func(span sdk.Span, headers map[string][]string) bool
-	BodyEvaluators    []func(span sdk.Span, body []byte) bool
-}
-
-// EvaluateURL runs URL evaluators in the filter
-func (f Filter) EvaluateURL(span sdk.Span, url string) bool {
-	for _, urlEvaluator := range f.URLEvaluators {
-		if urlEvaluator(span, url) {
+// Filter evaluates whether request should be blocked
+type Filter interface {
+	EvaluateURL(span sdk.Span, url string) bool
+	EvaluateHeaders(span sdk.Span, headers map[string][]string) bool
+	EvaluateBody(span sdk.Span, body []byte) bool
```

Comment:
I still believe we should use struct with null checks rather than an interface, otherwise we will read the body before the userland code no matter the user intent.

Can't we do 

```go
type Filter struct {
    EvaluateURL func (span sdk.Span, url string) bool
    EvaluateHeaders func (span sdk.Span, headers map[string][]string) bool
    EvaluateBody func (span sdk.Span, body []byte) bool
}
```

In addition, consider using the headerAccessor for headers as that is what we use for HTTP (cc https://github.com/hypertrace/goagent/pull/87)

Reply:
I think we want consistency in interface with Java Agent's. And we're ok with tying body filtering with data capture. See here: https://github.com/hypertrace/goagent/pull/83#discussion_r578839469.

Right, I looked at headerAccessor but that's specific to HTTP instrumentation, whereas we want filter to be generic for other instrumentations e.g. grpc.